src/xrScriptEngine/script_engine.cpp

```
@@ -754,8 +753,20 @@ void CScriptEngine::lua_error(lua_State* L)
 
 int CScriptEngine::lua_pcall_failed(lua_State* L)
 {
-    print_output(L, "", LUA_ERRRUN);
+#if (!defined(DEBUG) && !XRAY_EXCEPTIONS)
+    print_output(L, "", 0); // Force game to not break in raii_guard() for this type of errors
+#else
+    print_output(L, "", LUA_ERRRUN); // Default behavior
+#endif
```

I discussed this with @nitrocaster, we're more in favor of breaking the game on any type of error even in shipping code. We don't want scripts to silently fail.

---

It's break any way on any type of error. Just in another place.

FATAL ERROR
[error] Expression    : <no expression>
**[error] Function      : CScriptEngine::lua_pcall_failed
[error] File          : script_engine.cpp**

---

That's still a bit inconsistent, why should only 10% break in raii_guard()?

---

.___. it's all working, just belive me or test it yourself in game.

---

@Armada651 That's how it is originally designed (it has to be fixed, but that's another story). So, are we good now?

---

@ShokerStlk If `lua_pcall_failed` doesn't crash the game, what's going to crash it? We're trying to understand why it is so that we can not handle all lua errors in the same way.

---

Did i say somewhere that lua_pcall_failed  didn't crash the game? There is ::Fatal() right at the end of lua_pcall_failed body.

---

Okay, i have some idea.

---

OK, here's what I got so far:

- `!XRAY_EXCEPTIONS` is effectively equal to `!defined(DEBUG)`.
- luabind uses `pcall()` to execute lua functions. Each time `pcall()` is called, pcall_callback is passed to luajit by luabind. When lua function fails, luajit calls pcall_callback.
- When `pcall()` fails (e.g. lua function fails), luabind calls `luabind::detail::call_error()`, which in it's turn calls error callback (`CScriptEngine::lua_error()`) followed by `assert()` and `std::terminate()`. So, in release configuration the crash occurs in `CScriptEngine::lua_error()`, and in debug configuration it happens in `CScriptEngine::lua_pcall_failed()`. Even when neither of these two callbacks crashes the game, luabind crashes the game by calling `assert()` and `std::terminate()`.
- luabind does not use error callback (see `luabind::get_error_callback()`) in release, which is the exact reason why we can not handle all errors in `CScriptEngine::lua_error()`.

Hope that helps to understand this mess.